src/java.base/share/classes/java/lang/runtime/Carrier.java

```diff
@@ -331,8 +331,9 @@ private static MethodHandle reshapeConstructor(CarrierShape carrierShape,
              * @param objectCount     slot count required for objects
              */
             CarrierArray(int primitiveCount, int objectCount) {
-                this.primitives = new long[primitiveCount];
-                this.objects = new Object[objectCount];
+                this.primitives =
+                    primitiveCount != 0 ? new long[(primitiveCount + 1) / 2] : null;
```

Comment:
It might be better to use an unsigned shift here:
```suggestion
                    primitiveCount != 0 ? new long[(primitiveCount + 1) >>> 1] : null;
```

Reply:
doesn't the compiler (either javac or hotspot) automatically do that?

Reply:
Yes, the compilers handle this.  I did however switch to 

```
                    primitiveCount != 0 ? new long[(primitiveCount + 1) / LONG_SLOTS] : null;
```

Reply:
> doesn't the compiler (either javac or hotspot) automatically do that?

Only for multiplication, because `-1 >> 1 == -1`, `-1 >>> 1 == Integer.MAX_VALUE`, and `-1 / 2 == 0`.